File: backend/security_engine/risk_scorer.py
from typing import List, Dict, Any
from .analyzer import analyze_prompt
from .anonymizer import anonymize_prompt

def evaluate_security(original_prompt: str, policies: Dict[str, bool], custom_watchlist: List[str] = None) -> Dict[str, Any]:
    entities = analyze_prompt(original_prompt, custom_watchlist)

    raw_score = 0
    block_reasons = []
    matched_policies = []

    has_financial = False
    has_source_code = False
    has_secrets_or_ssn = False

    for entity in entities:
        t = entity["type"]
        if t == "API_KEY_SECRET":
            raw_score += 35
            has_secrets_or_ssn = True
        elif t == "US_SSN":
            raw_score += 30
            has_secrets_or_ssn = True
        elif t == "CREDIT_CARD":
            raw_score += 30
            has_secrets_or_ssn = True
        elif t == "FINANCIAL_VALUE":
            raw_score += 25
            has_financial = True
        elif t == "CLIENT_WATCHLIST":
            raw_score += 20
        elif t == "SOURCE_CODE_INDICATOR":
            raw_score += 15
            has_source_code = True
        elif t in ("EMAIL", "PHONE"):
            raw_score += 10

    risk_score = min(100, raw_score)
    action = "ALLOWED"

    if policies.get("block_financial", False) and has_financial:
        action = "BLOCKED"
        block_reasons.append("Contains prohibited financial values or revenue exposure.")
        matched_policies.append("block_financial")

    if policies.get("block_source_code", False) and has_source_code:
        action = "BLOCKED"
        block_reasons.append("Contains proprietary source code or system constructs.")
        matched_policies.append("block_source_code")

    if policies.get("strict_zero_trust", False) and (risk_score >= 30 or has_secrets_or_ssn):
        action = "BLOCKED"
        block_reasons.append("Strict Zero-Trust policy violated (risk score >= 30 or sensitive credentials).")
        matched_policies.append("strict_zero_trust")

    if risk_score >= 75 and action != "BLOCKED":
        action = "BLOCKED"
        block_reasons.append("Overall risk score exceeded maximum threshold (75/100).")
        matched_policies.append("max_risk_threshold")

    if action != "BLOCKED" and entities:
        if policies.get("redact_pii", True):
            action = "REDACTED"
            matched_policies.append("redact_pii")

    processed_prompt = original_prompt
    if action == "BLOCKED":
        processed_prompt = f"[PROMPT BLOCKED BY AEGIS GATEWAY POLICY: {' '.join(block_reasons)}]"
    elif action == "REDACTED":
        processed_prompt = anonymize_prompt(original_prompt, entities, policies.get("redact_pii", True))

    return {
        "original_prompt": original_prompt,
        "processed_prompt": processed_prompt,
        "action": action,
        "risk_score": risk_score,
        "entities": entities,
        "matched_policies": matched_policies,
        "block_reasons": block_reasons
    }

File: src/security_engine/risk_scorer.ts
import { EntityMatch, PolicyAction, SecurityAnalysisResult } from '../types';
import { analyzePrompt } from './analyzer';
import { anonymizePrompt } from './anonymizer';

export interface PolicyState {
  redact_pii: boolean;
  block_financial: boolean;
  block_source_code: boolean;
  strict_zero_trust: boolean;
}

/**
 * Calculates a 0-100 risk score and enforces policy actions.
 */
export function evaluateSecurity(
  originalPrompt: string,
  policies: PolicyState,
  customWatchlist?: string[]
): SecurityAnalysisResult {
  const entities: EntityMatch[] = analyzePrompt(originalPrompt, customWatchlist);

  let rawScore = 0;
  const blockReasons: string[] = [];
  const matchedPolicies: string[] = [];

  let hasFinancial = false;
  let hasSourceCode = false;
  let hasSecretsOrSsn = false;

  for (const entity of entities) {
    switch (entity.type) {
      case 'API_KEY_SECRET':
        rawScore += 35;
        hasSecretsOrSsn = true;
        break;
      case 'US_SSN':
        rawScore += 30;
        hasSecretsOrSsn = true;
        break;
      case 'CREDIT_CARD':
        rawScore += 30;
        hasSecretsOrSsn = true;
        break;
      case 'FINANCIAL_VALUE':
        rawScore += 25;
        hasFinancial = true;
        break;
      case 'CLIENT_WATCHLIST':
        rawScore += 20;
        break;
      case 'SOURCE_CODE_INDICATOR':
        rawScore += 15;
        hasSourceCode = true;
        break;
      case 'EMAIL':
        rawScore += 10;
        break;
      case 'PHONE':
        rawScore += 10;
        break;
    }
  }

  const riskScore = Math.min(100, rawScore);

  // Determine policy action
  let action: PolicyAction = 'ALLOWED';

  // Check blocking conditions
  if (policies.block_financial && hasFinancial) {
    action = 'BLOCKED';
    blockReasons.push('Contains prohibited financial values or revenue exposure.');
    matchedPolicies.push('block_financial');
  }

  if (policies.block_source_code && hasSourceCode) {
    action = 'BLOCKED';
    blockReasons.push('Contains proprietary source code or system constructs.');
    matchedPolicies.push('block_source_code');
  }

  if (policies.strict_zero_trust && (riskScore >= 30 || hasSecretsOrSsn)) {
    action = 'BLOCKED';
    blockReasons.push('Strict Zero-Trust policy violated (risk score >= 30 or sensitive credentials).');
    matchedPolicies.push('strict_zero_trust');
  }

  if (riskScore >= 75 && action !== 'BLOCKED') {
    action = 'BLOCKED';
    blockReasons.push('Overall risk score exceeded maximum threshold (75/100).');
    matchedPolicies.push('max_risk_threshold');
  }

  // If not blocked, check if redacted
  if (action !== 'BLOCKED' && entities.length > 0) {
    if (policies.redact_pii) {
      action = 'REDACTED';
      matchedPolicies.push('redact_pii');
    }
  }

  // Always produce the refined/anonymized version for processedPrompt
  let processedPrompt = anonymizePrompt(originalPrompt, entities, true);

  return {
    original_prompt: originalPrompt,
    processed_prompt: processedPrompt,
    action,
    risk_score: riskScore,
    entities,
    matched_policies: matchedPolicies,
    block_reasons: blockReasons
  };
}
